src/annotator/anchoring/text-range.js

```diff
@@ -0,0 +1,169 @@
+ *
+ * This position can be resolved to a specific descendant node in the current
+ * DOM subtree of the element using the `resolve` method.
+ */
+export class TextPosition {
+  /**
+   * Construct a `TextPosition` that refers to the text position `offset` within
+   * the text content of `element`.
+   *
+   * @param {Element} element
+   * @param {number} offset
+   */
+  constructor(element, offset) {
+    if (offset < 0) {
+      throw new Error('Offset is invalid');
+    }
+
+    this.element = element;
+
+    /** Character offset from the start of the element's `textContent`. */
+    this.offset = offset;
+  }
+
+  /**
+   * Resolve the position to a specific text node and offset within that node.
+   *
+   * Throws if `this.offset` exceeds the length of the element's text or if
+   * the element has no text. Offsets at the boundary between two nodes are
+   * resolved to the start of the node that begins at the boundary.
+   *
+   * @return {{ node: Text, offset: number }}
+   * @throws {RangeError}
+   */
+  resolve() {
+    const root = this.element;
+    const nodeIter = /** @type {Document} */ (root.ownerDocument).createNodeIterator(
+      root,
+      NodeFilter.SHOW_TEXT
+    );
+
+    let currentNode;
+    let textNode;
+    let length = 0;
+
+    while ((currentNode = nodeIter.nextNode())) {
+      textNode = /** @type {Text} */ (currentNode);
+      if (length + textNode.data.length > this.offset) {
+        return { node: textNode, offset: this.offset - length };
+      }
+      length += textNode.data.length;
+    }
+
+    if (textNode && length === this.offset) {
+      return { node: textNode, offset: textNode.data.length };
+    }
+
+    throw new RangeError('Offset exceeds text length');
+  }
+
+  /**
+   * Construct a `TextPosition` representing the range start or end point (node, offset).
+   *
+   * @param {Node} node
+   * @param {number} offset
+   * @return {TextPosition}
+   */
+  static fromPoint(node, offset) {
+    switch (node.nodeType) {
+      case Node.TEXT_NODE: {
+        if (offset < 0 || offset > /** @type {Text} */ (node).data.length) {
+          throw new Error('Text node offset is out of range');
+        }
+
+        if (!node.parentElement) {
+          throw new Error('Text node has no parent');
+        }
+
+        // Get the offset to the start of the parent element.
+        const textOffset = previousSiblingTextLength(node) + offset;
+
+        return new TextPosition(node.parentElement, textOffset);
+      }
+      case Node.ELEMENT_NODE: {
+        if (offset < 0 || offset > node.childNodes.length) {
+          throw new Error('Child node offset is out of range');
+        }
+
+        // Get the text length before the `offset`th child of element.
+        let textOffset = 0;
+        for (let i = 0; i < offset; i++) {
+          textOffset += node.childNodes[i].textContent?.length ?? 0;
+        }
+
+        return new TextPosition(/** @type {Element} */ (node), textOffset);
```

Comment:
`TextPosition`'s constructor requires an `Element`. This hint is needed because TypeScript is not smart enough to realize that the enclosing `switch` statement implies the node is of a certain type.